core/src/main/java/org/apache/spark/shuffle/sort/io/DefaultShuffleExecutorComponents.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.shuffle.sort.io;
+
+import org.apache.spark.SparkConf;
+import org.apache.spark.SparkEnv;
+import org.apache.spark.TaskContext;
+import org.apache.spark.api.shuffle.ShuffleExecutorComponents;
+import org.apache.spark.api.shuffle.ShuffleWriteSupport;
+import org.apache.spark.executor.TaskMetrics;
+import org.apache.spark.shuffle.IndexShuffleBlockResolver;
+import org.apache.spark.storage.BlockManager;
+
+public class DefaultShuffleExecutorComponents implements ShuffleExecutorComponents {
+
+  private final SparkConf sparkConf;
+  private BlockManager blockManager;
+  private IndexShuffleBlockResolver blockResolver;
+  private TaskMetrics metrics;
+
+  public DefaultShuffleExecutorComponents(SparkConf sparkConf) {
+    this.sparkConf = sparkConf;
+  }
+
+  @Override
+  public void intitializeExecutor(String appId, String execId) {
+    blockManager = SparkEnv.get().blockManager();
+    blockResolver = new IndexShuffleBlockResolver(sparkConf, blockManager);
+    metrics = TaskContext.get().taskMetrics();
```

Comment:
wait, i don't think you want this here actually. The TaskContext is associated with each individual shuffle task, but you're only calling `initializeExecutor` once per executor. You want to get the metrics from `TaskContext` after you have the set the taskContext here (https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/scheduler/Task.scala#L102) for each task

Reply:
@mccheah if so, can we pass `metrics` into the ShuffleMapOutputWriter so that it can be properly mocked (can't mock Static methods).

Reply:
There's a few ways you can do it. You can pass it in as you suggested. In this particular case, you can also call `TaskContext.set(mockTaskContext)` to initialize the static variable before the test. I prefer the latter way, but you could make a case for either.

Reply:
Actually you might want to pass it through so that concurrent tests don't collide...

Reply:
The problem is that `TaskMetrics` is currently marked as `DeveloperApi` which means it's questionable to pass it in to a public API. We could proposed some alternative metrics API that delegates to the Spark default `TaskMetrics` API. But I think for now we can use `TaskContext#get` from inside the writer and then in tests call `TaskContext#setTaskContext`. I took a closer look and a lot of tests use `TaskContext#setTaskContext` indicating they don't anticipate tests to be run in parallel in the same JVM.

Reply:
Just make sure we call `TaskContext#unset` appropriate after each test.